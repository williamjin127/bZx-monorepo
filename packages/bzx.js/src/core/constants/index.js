import { BigNumber } from "bignumber.js";

export const SOLIDITY_TYPE_MAX_CHARS = 64;
export const NULL_ADDRESS = "0x0000000000000000000000000000000000000000";
export const UNLIMITED_ALLOWANCE_IN_BASE_UNITS = new BigNumber(2)
  .pow(256)
  .minus(1);
export const MAX_DIGITS_IN_UNSIGNED_256_INT = 78;
