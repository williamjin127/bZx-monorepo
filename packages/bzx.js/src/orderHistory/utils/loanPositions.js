import { pipe, map } from "ramda";
import * as Utils from "./index";

const NUM_LOAN_POS_FIELDS = 15;

const getLoanPosition = params => ({
  ...map(pipe(Utils.substr24, Utils.prepend0x), {
    lender: params[0],
    trader: params[1],
    collateralTokenAddressFilled: params[2],
    positionTokenAddressFilled: params[3],
    loanTokenAddress: params[11],
    interestTokenAddress: params[12]
  }),
  ...map(pipe(Utils.prepend0x, Utils.parseIntHex), {
    loanTokenAmountFilled: params[4],
    collateralTokenAmountFilled: params[5],
    positionTokenAmountFilled: params[6],
    loanStartUnixTimestampSec: params[7],
    loanEndUnixTimestampSec: params[8],
    active: params[9],
    interestPaidTotal: params[13],
    interestDepositRemaining: params[14]
  }),
  ...map(pipe(Utils.prepend0x), {
    loanOrderHash: params[10]
  })
});

const checkProperObjCount = Utils.makeCheckProperObjCount(NUM_LOAN_POS_FIELDS);
const getOrderObjArray = Utils.makeGetOrderObjArray(NUM_LOAN_POS_FIELDS);

export const cleanData = raw =>
  raw
    ? pipe(
        Utils.remove0xPrefix,
        checkProperObjCount,
        getOrderObjArray,
        map(
          pipe(
            Utils.getOrderParams,
            getLoanPosition
          )
        )
      )(raw)
    : [];
