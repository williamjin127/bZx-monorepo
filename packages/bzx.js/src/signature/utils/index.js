import * as ethUtil from "ethereumjs-util";

export const isValidSignature = (data, signature, signerAddress) => {
  const dataBuff = ethUtil.toBuffer(data);
  const msgHashBuff = ethUtil.hashPersonalMessage(dataBuff);
  try {
    const pubKey = ethUtil.ecrecover(
      msgHashBuff,
      signature.v,
      ethUtil.toBuffer(signature.r),
      ethUtil.toBuffer(signature.s)
    );
    const retrievedAddress = ethUtil.bufferToHex(ethUtil.pubToAddress(pubKey));
    return retrievedAddress === signerAddress;
  } catch (err) {
    return false;
  }
};

export const parseSignatureHexAsVRS = signatureHex => {
  const signatureBuffer = ethUtil.toBuffer(signatureHex);
  let v = signatureBuffer[0];
  if (v < 27) {
    v += 27;
  }
  const r = signatureBuffer.slice(1, 33);
  const s = signatureBuffer.slice(33, 65);
  const ecSignature = {
    v,
    r: ethUtil.bufferToHex(r),
    s: ethUtil.bufferToHex(s)
  };
  return ecSignature;
};

export const parseSignatureHexAsRSV = signatureHex => {
  const sig = ethUtil.fromRpcSig(signatureHex);
  const ecSignature = {
    v: sig.v,
    r: ethUtil.bufferToHex(sig.r),
    s: ethUtil.bufferToHex(sig.s)
  };
  return ecSignature;
};
