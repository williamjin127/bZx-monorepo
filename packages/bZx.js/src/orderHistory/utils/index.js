import { SOLIDITY_TYPE_MAX_CHARS } from "../../core/constants";

export const remove0xPrefix = data => (data ? data.substr(2) : "");

export const makeCheckProperObjCount = numFields => data => {
  const objCount = data.length / SOLIDITY_TYPE_MAX_CHARS / numFields;
  if (objCount % 1 !== 0)
    throw new Error("Data length invalid, must be whole number of objects");
  return data;
};

export const makeGetOrderObjArray = numFields => data =>
  data.match(new RegExp(`.{1,${numFields * SOLIDITY_TYPE_MAX_CHARS}}`, "g"));

export const getOrderParams = data =>
  data.match(new RegExp(`.{1,${SOLIDITY_TYPE_MAX_CHARS}}`, "g"));

const HEX_RADIX = 16;
export const prepend0x = arg => `0x${arg}`;
export const substr24 = arg => arg.substr(24);
export const parseIntHex = arg => parseInt(arg, HEX_RADIX);
