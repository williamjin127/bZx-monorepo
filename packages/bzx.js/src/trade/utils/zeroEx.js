import * as constants from "../../core/constants";

export const transform0xOrder = ({ signedOrder }) => ({
  exchangeContractAddress: signedOrder.exchangeContractAddress,
  expirationUnixTimestampSec: signedOrder.expirationUnixTimestampSec,
  feeRecipient: signedOrder.feeRecipient,
  maker: signedOrder.maker,
  makerFee: signedOrder.makerFee,
  makerTokenAddress: signedOrder.makerTokenAddress,
  makerTokenAmount: signedOrder.makerTokenAmount,
  salt: signedOrder.salt,
  taker: signedOrder.taker || constants.NULL_ADDRESS,
  takerFee: signedOrder.takerFee,
  takerTokenAddress: signedOrder.takerTokenAddress,
  takerTokenAmount: signedOrder.takerTokenAmount
});
