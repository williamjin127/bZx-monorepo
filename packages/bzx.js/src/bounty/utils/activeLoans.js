import { pipe, map } from "ramda";
import * as OrderHistoryUtils from "../../orderHistory/utils/index";

const NUM_LOAN_FIELDS = 3;

const getLoan = params => ({
  loanOrderHash: OrderHistoryUtils.prepend0x(params[0]),
  trader: pipe(
    OrderHistoryUtils.substr24,
    OrderHistoryUtils.prepend0x
  )(params[1]),
  loanEndUnixTimestampSec: OrderHistoryUtils.parseIntHex(params[2])
});

const checkProperObjCount = OrderHistoryUtils.makeCheckProperObjCount(
  NUM_LOAN_FIELDS
);
const getOrderObjArray = OrderHistoryUtils.makeGetOrderObjArray(
  NUM_LOAN_FIELDS
);

export const cleanData = raw =>
  raw
    ? pipe(
        OrderHistoryUtils.remove0xPrefix,
        checkProperObjCount,
        getOrderObjArray,
        map(
          pipe(
            OrderHistoryUtils.getOrderParams,
            getLoan
          )
        )
      )(raw)
    : [];
