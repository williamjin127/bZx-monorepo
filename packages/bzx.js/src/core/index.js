import { assert } from "@0xproject/assert";
import { BigNumber } from "@0xproject/utils";
import * as constants from "./constants";
import { schemas } from "../schemas/bZx_json_schemas";
import * as utils from "./utils";
import * as tokenRegistry from "../tokenRegistry";
import EIP20 from "../contracts/EIP20.json";
import * as allowance from "../allowance";
import * as oracles from "../oracles";
import * as fill from "../fill";
import * as Addresses from "../addresses";
import * as orderHistory from "../orderHistory";
import * as transfer from "../transfer";
import * as signature from "../signature";
import * as Errors from "./constants/errors";
import * as trade from "../trade";
import * as loanHealth from "../loanHealth";
import * as bounty from "../bounty";
import * as weth from "../weth";

export class BZxJS {
  static generatePseudoRandomSalt = utils.generatePseudoRandomSalt;
  static noop = utils.noop;
  static toChecksumAddress = utils.toChecksumAddress;

  /* On Metamask, provider.host is undefined
  Force users to provide host url */
  constructor(web3, { networkId, addresses = Addresses.getAddresses(networkId) } = {}) {
    if (!networkId) throw new Error(Errors.NoNetworkId);

    this.web3 = web3;
    this.addresses = addresses;
    this.networkId = networkId;
    switch (networkId) {
      case 1:
        this.networkName = "mainnet";
        this.etherscanURL = "https://etherscan.io/";
        break;
      case 3:
        this.networkName = "ropsten";
        this.etherscanURL = "https://ropsten.etherscan.io/";
        break;
      case 4:
        this.networkName = "rinkeby";
        this.etherscanURL = "https://rinkeby.etherscan.io/";
        break;
      case 42:
        this.networkName = "kovan";
        this.etherscanURL = "https://kovan.etherscan.io/";
        break;
      default:
        this.networkName = "local";
        this.etherscanURL = "";
        break;
    }
  }

  static getLoanOrderHashHex(order) {
    utils.doesConformToSchema("loanOrder", order, schemas.loanOrderSchema);
    const orderHashHex = utils.getLoanOrderHashHex(order);
    return orderHashHex;
  }
  getLoanOrderHashAsync = async props =>
    utils.getLoanOrderHashAsync(this, props);

  static isValidSignature = props => signature.isValidSignature(props);

  isValidSignatureAsync = async props =>
    signature.isValidSignatureAsync(this, props);

  signOrderHashAsync = async (...props) =>
    signature.signOrderHashAsync(this, ...props);

  setAllowance = (...props) => allowance.setAllowance(this, ...props);

  setAllowanceUnlimited = props =>
    this.setAllowance({
      ...props,
      amountInBaseUnits: constants.UNLIMITED_ALLOWANCE_IN_BASE_UNITS
    });

  resetAllowance = props =>
    this.setAllowance({
      ...props,
      amountInBaseUnits: new BigNumber(0)
    });

  getAllowance = async (...props) => allowance.getAllowance(this, ...props);

  getBalance = async ({ tokenAddress, ownerAddress }) => {
    assert.isETHAddressHex("ownerAddress", ownerAddress);
    assert.isETHAddressHex("tokenAddress", tokenAddress);

    const tokenContract = await utils.getContractInstance(
      this.web3,
      EIP20.abi,
      tokenAddress
    );
    const balance = await tokenContract.methods.balanceOf(ownerAddress).call();
    return new BigNumber(balance);
  };

  getTokenList = async () => tokenRegistry.getTokenList(this);

  getOracleList = async () => oracles.getOracleList(this);
  isTradeSupported = async (...props) =>
    oracles.isTradeSupported(this, ...props);
  getConversionData = async (...props) =>
    oracles.getConversionData(this, ...props);

  takeLoanOrderAsLender = (...props) =>
    fill.takeLoanOrderAsLender(this, ...props);

  takeLoanOrderAsTrader = (...props) =>
    fill.takeLoanOrderAsTrader(this, ...props);

  pushLoanOrderOnChain = (...props) =>
    fill.pushLoanOrderOnChain(this, ...props);

  takeLoanOrderOnChainAsTrader = (...props) =>
    fill.takeLoanOrderOnChainAsTrader(this, ...props);

  takeLoanOrderOnChainAsLender = (...props) =>
    fill.takeLoanOrderOnChainAsLender(this, ...props);

  cancelLoanOrder = (...props) => fill.cancelLoanOrder(this, ...props);

  cancelLoanOrderWithHash = (...props) =>
    fill.cancelLoanOrderWithHash(this, ...props);

  getInitialCollateralRequired = async (...props) =>
    fill.getInitialCollateralRequired(this, ...props);

  orderFilledAmount = async (...props) =>
    fill.orderFilledAmount(this, ...props);
  orderCancelledAmount = async (...props) =>
    fill.orderCancelledAmount(this, ...props);

  getSingleOrder = async (...props) =>
    orderHistory.getSingleOrder(this, ...props);
  getOrdersFillable = async (...props) =>
    orderHistory.getOrdersFillable(this, ...props);
  getOrdersForUser = async (...props) =>
    orderHistory.getOrdersForUser(this, ...props);
  getSingleLoan = async (...props) =>
    orderHistory.getSingleLoan(this, ...props);
  getLoansForLender = async (...props) =>
    orderHistory.getLoansForLender(this, ...props);
  getLoansForTrader = async (...props) =>
    orderHistory.getLoansForTrader(this, ...props);

  transferToken = (...props) => transfer.transferToken(this, ...props);

  tradePositionWith0x = (...props) => trade.tradePositionWith0x(this, ...props);
  tradePositionWith0xV2 = (...props) => trade.tradePositionWith0xV2(this, ...props);
  tradePositionWithOracle = (...props) =>
    trade.tradePositionWithOracle(this, ...props);

  depositCollateral = (...props) => loanHealth.depositCollateral(this, ...props);
  withdrawCollateral = (...props) => loanHealth.withdrawCollateral(this, ...props);
  changeCollateral = (...props) => loanHealth.changeCollateral(this, ...props);
  withdrawPosition = (...props) => loanHealth.withdrawPosition(this, ...props);
  depositPosition = (...props) => loanHealth.depositPosition(this, ...props);
  getPositionOffset = (...props) => loanHealth.getPositionOffset(this, ...props);

  closeLoanPartially = (...props) => loanHealth.closeLoanPartially(this, ...props);
  closeLoan = (...props) => loanHealth.closeLoan(this, ...props);

  payInterestForOracle = (...props) => loanHealth.payInterestForOracle(this, ...props);
  getLenderInterestForOracle = (...props) => loanHealth.getLenderInterestForOracle(this, ...props);
  getTraderInterestForLoan = (...props) => loanHealth.getTraderInterestForLoan(this, ...props);

  requestFaucetToken = (...props) => utils.requestFaucetToken(this, ...props);

  getActiveLoans = (...props) => bounty.getActiveLoans(this, ...props);
  getMarginLevels = (...props) => bounty.getMarginLevels(this, ...props);
  liquidateLoan = (...props) => bounty.liquidateLoan(this, ...props);

  wrapEth = (...props) => weth.wrapEth(this, ...props);
  unwrapEth = (...props) => weth.unwrapEth(this, ...props);

  getWeb3Contract = (...props) => utils.getWeb3Contract(this, ...props);
}

export default BZxJS;
