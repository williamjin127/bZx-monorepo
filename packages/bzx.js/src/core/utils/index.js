import { BigNumber } from "@0xproject/utils";
import { assert } from "@0xproject/assert";
import BN from "bn.js";
import Web3Utils from "web3-utils";
import * as constants from "../constants";
import { SchemaValidator } from "../../schemas/bZx_json_schemas";
import { getContracts } from "../../contracts";
import * as Addresses from "../../addresses";

export const noop = () => {};

export const bigNumberToBN = value => new BN(value.toString(), 10);

export const generatePseudoRandomSalt = () => {
  // BigNumber.random returns a pseudo-random number between 0 & 1
  // with a passed in number of decimal places.
  // Source: https://mikemcl.github.io/bignumber.js/#random
  const randomNumber = BigNumber.random(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT
  );
  const factor = new BigNumber(10).pow(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT - 1
  );
  const salt = randomNumber.times(factor).round();
  return salt;
};

const getOrderValues = (order, shouldFormatAsStrings) => {
  // Must be strings in production for Web3Utils.soliditySha3 for some reason
  if (shouldFormatAsStrings) {
    return [
      order.loanTokenAmount.toString(),
      order.interestAmount.toString(),
      order.initialMarginAmount.toString(),
      order.maintenanceMarginAmount.toString(),
      order.lenderRelayFee.toString(),
      order.traderRelayFee.toString(),
      order.maxDurationUnixTimestampSec.toString(),
      order.expirationUnixTimestampSec.toString(),
      order.makerRole.toString(),
      order.withdrawOnOpen.toString(),
      order.salt.toString()
    ];
  }
  return [
    bigNumberToBN(order.loanTokenAmount),
    bigNumberToBN(order.interestAmount),
    bigNumberToBN(order.initialMarginAmount),
    bigNumberToBN(order.maintenanceMarginAmount),
    bigNumberToBN(order.lenderRelayFee),
    bigNumberToBN(order.traderRelayFee),
    bigNumberToBN(order.maxDurationUnixTimestampSec),
    bigNumberToBN(order.expirationUnixTimestampSec),
    bigNumberToBN(order.makerRole),
    bigNumberToBN(order.withdrawOnOpen),
    bigNumberToBN(order.salt)
  ];
};

const getLoanOrderHashArgs = (order, shouldFormatAsStrings) => {
  const orderAddresses = [
    order.makerAddress,
    order.loanTokenAddress,
    order.interestTokenAddress,
    order.collateralTokenAddress,
    order.feeRecipientAddress,
    order.oracleAddress,
    order.takerAddress,
    order.tradeTokenToFillAddress
  ];
  const orderValues = getOrderValues(order, shouldFormatAsStrings);
  const oracleData = order.oracleData || "0x";

  return { orderAddresses, orderValues, oracleData };
};

export const doesContractExistAtAddress = async (web3, address) => {
  const code = await web3.eth.getCode(address);
  // Regex matches 0x0, 0x00, 0x in order to accommodate poorly implemented clients
  const codeIsEmpty = /^0x0{0,40}$/i.test(code);
  return !codeIsEmpty;
};

export const getContractInstance = (web3, abi, address) => {
  assert.isETHAddressHex("address", address);
  const contract = new web3.eth.Contract(abi, address);
  return contract;
};

export const getLoanOrderHashHex = order => {
  const { orderAddresses, orderValues, oracleData } = getLoanOrderHashArgs(order, true);
  const orderHashHex = Web3Utils.soliditySha3(
    { t: "address", v: order.bZxAddress },
    { t: "address[8]", v: orderAddresses },
    { t: "uint256[11]", v: orderValues },
    { t: "bytes", v: oracleData }
  );
  return orderHashHex;
};

export const getLoanOrderHashAsync = async ({ web3, networkId }, order) => {
  const { orderAddresses, orderValues, oracleData } = getLoanOrderHashArgs(order, true);
  const bZxContract = await getContractInstance(
    web3,
    getContracts(networkId).BZx.abi,
    Addresses.getAddresses(networkId).BZx
  );
  return bZxContract.methods
    .getLoanOrderHash(
      orderAddresses,
      orderValues,
      oracleData
    )
    .call();
};

export const doesConformToSchema = (variableName, value, schema) => {
  const schemaValidator = new SchemaValidator();
  const validationResult = schemaValidator.validate(value, schema);
  const hasValidationErrors = validationResult.errors.length > 0;
  const msg = `Expected ${variableName} to conform to schema ${
    schema.id
  }\nEncountered: ${JSON.stringify(
    value,
    null,
    "\t"
  )}\nValidation errors: ${validationResult.errors.join(", ")}`;
  assert.assert(!hasValidationErrors, msg);
};

export const toChecksumAddress = addr => Web3Utils.toChecksumAddress(addr);

export const requestFaucetToken = (
  { web3, networkId },
  { tokenAddress, receiverAddress, getObject, txOpts }
) => {
  const faucetContract = getContractInstance(
    web3,
    getContracts(networkId).TestNetFaucet.abi,
    Addresses.getAddresses(networkId).TestNetFaucet
  );

  const txObj = faucetContract.methods.faucet(
    toChecksumAddress(tokenAddress),
    toChecksumAddress(receiverAddress)
  );
  console.log(`requestFaucetToken: ${txObj.encodeABI()}`);

  if (getObject) {
    return txObj;
  }
  return txObj.send(txOpts);
};

export const getWeb3Contract = ({ web3, networkId }, contractName, contractAddress) => getContractInstance(
    web3,
    getContracts(networkId)[contractName].abi,
    contractAddress ? contractAddress : Addresses.getAddresses(networkId)[contractName]
  );
