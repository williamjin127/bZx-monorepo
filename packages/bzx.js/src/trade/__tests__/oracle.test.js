import { pathOr, pipe } from "ramda";
import { BZxJS } from "../../core/index";
import bZxJS from "../../core/__tests__/setup";
import * as FillTestUtils from "../../fill/__tests__/utils";
import * as CoreTestUtils from "../../core/__tests__/utils";

describe("trade", () => {
  const { web3 } = bZxJS;
  const {
    loanTokens,
    interestTokens,
    collateralTokens
  } = FillTestUtils.initAllContractInstances();

  const { owner, lenders, traders } = FillTestUtils.getAccounts();

  const takerAddress = traders[0];
  const txOpts = {
    from: takerAddress,
    gas: 1000000,
    gasPrice: web3.utils.toWei("5", "gwei").toString()
  };

  let orderHash = null;
  let promiEvent = null;
  let takeLoanOrderAsTraderReceipt = null;

  beforeAll(async () => {
    await FillTestUtils.setupAll({
      owner,
      lenders,
      traders,
      transferAmount: web3.utils.toWei("1000000", "ether")
    });

    const order = FillTestUtils.makeOrderAsLender({
      web3,
      lenders,
      loanTokens,
      interestTokens
    });
    orderHash = BZxJS.getLoanOrderHashHex(order);

    // bZx hash that we give to tradePositionWith0x must belong to a loan that was previously filled, so we fill the loan order here
    const signature = await bZxJS.signOrderHashAsync(
      orderHash,
      order.makerAddress
    );
    const loanTokenAmountFilled = web3.utils.toWei("12.3");
    takeLoanOrderAsTraderReceipt = await bZxJS.takeLoanOrderAsTrader(
      { ...order, signature },
      collateralTokens[0].options.address.toLowerCase(),
      loanTokenAmountFilled,
      txOpts
    );

    promiEvent = bZxJS.tradePositionWithOracle({
      orderHash,
      tradeTokenAddress: interestTokens[1].options.address.toLowerCase(),
      txOpts
    });
  });

  describe("tradePositionWithOracle", () => {
    test("should return a web3 PromiEvent", async () => {
      CoreTestUtils.expectPromiEvent(promiEvent);
    });

    test("should successfully trade position with oracle (Kyber)", async () => {
      expect(
        pathOr(null, ["events", "DebugLine"], takeLoanOrderAsTraderReceipt)
      ).toBe(null);

      const receipt = await promiEvent;
      const debugLine = pathOr(null, ["events", "DebugLine"], receipt);

      const replaceRandomFields = CoreTestUtils.makeReplaceRandomFields([
        "0",
        "loanOrderHash"
      ]);

      const logMarginLevels = pipe(
        pathOr(null, ["events", "LogMarginLevels", "returnValues"]),
        replaceRandomFields
      )(receipt);

      const logPositionTraded = pipe(
        pathOr(null, ["events", "LogPositionTraded", "returnValues"]),
        replaceRandomFields
      )(receipt);

      expect(debugLine).toEqual(null);
      expect(logMarginLevels).toMatchSnapshot();
      expect(logPositionTraded).toMatchSnapshot();
    });
  });
});
