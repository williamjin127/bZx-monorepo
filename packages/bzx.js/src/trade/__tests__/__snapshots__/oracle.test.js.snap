// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trade tradePositionWithOracle should successfully trade position with oracle (Kyber) 1`] = `
Object {
  "0": "this is a random field",
  "1": "0xE36Ea790bc9d7AB70C55260C66D52b1eca985f84",
  "2": "50",
  "3": "25",
  "4": "50000000000000000000",
  "currentMarginAmount": "50000000000000000000",
  "initialMarginAmount": "50000000000000000000",
  "loanOrderHash": "this is a random field",
  "maintenanceMarginAmount": "25000000000000000000",
  "trader": "0xE36Ea790bc9d7AB70C55260C66D52b1eca985f84",
}
`;

exports[`trade tradePositionWithOracle should successfully trade position with oracle (Kyber) 2`] = `
Object {
  "0": "this is a random field",
  "1": "0xE36Ea790bc9d7AB70C55260C66D52b1eca985f84",
  "2": "0xB48E1B16829C7f5Bd62B76cb878A6Bb1c4625D7A",
  "3": "0x72D5A2213bfE46dF9FbDa08E22f536aC6Ca8907e",
  "4": "12300000000000000000",
  "5": "12300000000000000000",
  "destTokenAddress": "0x72D5A2213bfE46dF9FbDa08E22f536aC6Ca8907e",
  "destTokenAmount": "12300000000000000000",
  "loanOrderHash": "this is a random field",
  "sourceTokenAddress": "0xB48E1B16829C7f5Bd62B76cb878A6Bb1c4625D7A",
  "sourceTokenAmount": "12300000000000000000",
  "trader": "0xE36Ea790bc9d7AB70C55260C66D52b1eca985f84",
}
`;
